core/src/main/scala/kafka/server/AbstractFetcherThread.scala

```diff
@@ -715,6 +727,58 @@ abstract class AbstractFetcherThread(name: String,
     }
   }
 
+  /**
+   * Handle a partition whose offset is out of range and return a new fetch offset.
+   */
+  protected def fetchOffsetAndTruncate(topicPartition: TopicPartition, topicId: Option[Uuid], currentLeaderEpoch: Int): PartitionFetchState = {
+    fetchOffsetAndApplyFun(topicPartition, topicId, currentLeaderEpoch,
+      (epoch, leaderLogStartOffset) => truncateFullyAndStartAt(topicPartition, leaderLogStartOffset))
```

Comment:
Sure, I will look into this more.

Reply:
We encounter `OffsetOutOfRangeException` only when the offsets are beyond the range of [logStartOffset, logEndOffset]. Why do we need to build remote log metadata in that case? I may be missing something here.

Reply:
Sure, as we discussed offline. I added the approach to keep OffsetOutOfRange like earlier, which is to fetch the log-start-offset and it may get a response of OffsetMovedToTieredStorage if it tries to fetch from log-start-offset and it is moved to tiered storage.